snorkel/candidates.py

```diff
@@ -1,24 +1,70 @@
-import sys
-from collections import defaultdict
 from itertools import chain
-from time import time
 from multiprocessing import Process, Queue, JoinableQueue
+from snorkel import SnorkelBase, SnorkelSession
+from parser import Context
+from sqlalchemy import Column, String, Integer, ForeignKey
+from sqlalchemy.orm import relationship, backref
+from sqlalchemy.types import PickleType
 from Queue import Empty
 
 
-class Candidate(object):
-    """A candidate object, **uniquely identified by its id**"""
-    def __init__(self, id):
-        self.id = id
-    
-    def __eq__(self, other):
-        try:
-            return self.id == other.id
-        except:
-            raise NotImplementedError()
-    
-    def __hash__(self):
-        return hash(self.id)
+QUEUE_COLLECT_TIMEOUT = 5
+
+
+class Candidates(SnorkelBase):
+    """A named collection of Candidate objects."""
+    __tablename__ = 'candidates'
+    id = Column(String, primary_key=True)
+
+    def __repr__(self):
+        return "Candidates" + str((self.id,))
+
+    def __iter__(self):
+        """Default iterator is over Candidate objects"""
+        for candidate in self.candidates:
+            yield candidate
+
+    def get_candidates(self):
+        return self.candidates
+
+    def get_candidate(self, candidate_id):
+        """Retrieve a Candidate by id"""
+        session = SnorkelSession.object_session(self)
```

Comment:
This shouldn't actually create a new Session, it just returns whichever session the Candidates object belongs to.